### ISSUE TITLE
Split Serializers into Parsers and Serializers

### ISSUE THREAD

**Issue by davewasmer** [denali-js/denali#277]
Allow parsers to determine shape of object handed to responder method. Responder methods should take an object with `body`, `query` properties at a minimum, and parser can add more.

@acorncom @acburdine did I miss anything from our conversation?

**Comment by knownasilya** [comment]
Would love to see an example of this.

**Comment by davewasmer** [comment]
There's a few things this accomplishes:

### Splits the parser and serializer so a single action can have different classes for each

Right now, the serializer is chosen by the Model class to be rendered, or the Action can override it, or else it falls back to the ApplicationSerializer. When parsing the incoming request body, we similarly need to pick the right Serializer class to parse it, but that is trickier - we can't know the Model type of the payload (if it's even a Model) beforehand. So the default is to use the Action's override, or else fall back to the ApplicationSerializer. 

But lets take a `LoginAction` as an example - perhaps the incoming payload is a simple `{ email, password }`, but the response is a `Session` model that should be formatted according to JSON-API. Right now, you'd have to create a one-off Serializer class just for this `LoginAction` that does `FlatSerializer` parsing but `JSONAPISerializer` serializing. This change would let you simply say `parser = 'flat'` and `serializer = 'json-api'`.

```ts
// app/actions/login.ts
class LoginAction extends Action {

  parser = 'login';

  respond({ body }) {
    // body.email, body.password
    // ...
    return new Session();
  }
}
```

In the above example, the parser is explicitly specified (which is the only way to change it from the default ApplicationParser). The serializer will be picked based on the returned Model (in this case, SessionSerializer, falling back to ApplicationSerializer if not present).

### Parsers determine responder arguments shape

Right now, the standard app blueprint generates with the JSON-API serializer as the application default. But the standard resource blueprint generates with create and update actions that don't work with JSON-API:

```
respond(params) {
  let Post = this.modelFor('post');
  let post = Post.create(params); // <-- works with flat serializer, but JSON-API would be `params.data.attributes`
```

The example here shows how the JSON-API spec nests the typically-used inbound data inside `data.attributes`. Other common wire formats use similar nesting. One approach with our current setup would be to modify the JSON-API spec to simply "unwrap" the nested `data.attributes` object. But this is problematic, because it essentially "throws away" the rest of the payload (i.e. `included`, `meta`, `links`, etc).

This is fundamentally tied to the fact that whatever the Serializer.parse method returns is used to overwrite the `request.body`. The Serializer can't actually return any additional _structured_ data.

By changing the responder method signature from `respond(params)` (where `params` is the merged request body, parsed query params object, and URL fragments) to a structured object, we can allow the Serializer.parse method to return an object with that additional structured data:

```
// JsonApiSerializer.parse
parse(rawRequest) {
  return {
    body: rawRequest.body.data.attributes,
    query: keysToCamelCase(rawRequest.query),
    included: rawRequest.body.included,
    meta: rawRequest.body.meta
  };
}

// Action.respond
respond({ body }) {
  let Post = this.modelFor('post');
  let post = Post.create(body); // <-- works with both 'flat' and 'json-api' serializers now
```

Thus we can maintain JSON-API spec compliance (you can get to those `included` records in your responder method), while ensuring a uniform way of accessing the primary payload.

We'll likely include some basic assertions to ensure that the Serializer/Parser returns an object with at least `{ body, query, headers }` or something like that, to ensure some consistency on the common stuff.

**Comment by davewasmer** [comment]
BTW - feedback is welcome on this. If anyone sees a better approach we're missing, by all means, chime in ;)

**Comment by davewasmer** [comment]
fixed in 0133631872b46928304d5bd311f28fdfe774940b
